lib/src/main/java/io/ably/lib/realtime/Channel.java

```diff
@@ -215,6 +228,155 @@ public void sync() throws AblyException {
 	 * internal
 	 *
 	 */
+	/* Timer for attach operation */
+	private Timer attachTimer;
+
+	/* Timer for reattaching if attach failed */
+	private Timer reattachTimer;
+
+	/**
+	 * Cancel attach/reattach timers
+	 */
+	synchronized private void clearAttachTimers() {
+		Timer[] timers = new Timer[]{attachTimer, reattachTimer};
+		attachTimer = reattachTimer = null;
+		for (Timer t: timers) {
+			if (t != null) {
+				t.cancel();
+				t.purge();
+			}
+		}
+	}
+
+	/**
+	 * Attach channel, if not attached within timeout set state to suspended and
+	 * set up timer to reattach it later
+	 */
+	synchronized private void attachWithTimeout(final CompletionListener listener) throws AblyException {
+		final Timer currentAttachTimer = new Timer();
+		attachTimer = currentAttachTimer;
+
+		try {
+			attachImpl(new CompletionListener() {
+				@Override
+				public void onSuccess() {
+					clearAttachTimers();
+					if (listener != null)
+						listener.onSuccess();
+				}
+
+				@Override
+				public void onError(ErrorInfo reason) {
+					clearAttachTimers();
+					if (listener != null)
+						listener.onError(reason);
+				}
+			});
+		} catch(AblyException e) {
+			attachTimer = null;
+		}
+
+		if(attachTimer == null)
+			/* operation has already succeeded or failed, no need to set the timer */
+			return;
+
+		attachTimer.schedule(
+				new TimerTask() {
+					@Override
+					public void run() {
+						String errorMessage = String.format("Attach timed out for channel %s", name);
+						Log.v(TAG, errorMessage);
+						synchronized (Channel.this) {
+							if(attachTimer != currentAttachTimer)
+								return;
+							attachTimer = null;
+							if(state == ChannelState.attaching) {
+								setSuspended(new ErrorInfo(errorMessage, 91200));
+								reattachAfterTimeout();
```

Comment:
> I don't think the idea of the 0.8 spec is to forbid it, and ably-js has been doing these retries for long before it was put into the 0.9 spec.

There's a difference though as far as I can tell in that this moves the channel to `detached` with the error, and then retries.  If the channel remained in `attaching` and was retried that's a different thing.

Also, saying "I don't think the idea of the 0.8 spec is to forbid it" doesn't hold water for me.  The idea behind the spec is that there is predictable behaviour for customers.  The spec does not explicitly state all the possible things the library could do that it shouldn't do, it merely states what should happen so that developers understand expected behaviour.  If every dev working on the client libs added in behaviours they thought were useful but not part of the spec we'd have inconsistent libraries.